North sea example

jwallwork23 opened the pull request:
This PR adds an example of tidal modelling in the North Sea. I am planning to put detailed instructions of how to set up such a simulation in `examples/north_sea` and then add a shorter version with some pictures in `demos`, so that it will be visible on the website.

[Work in progress.]

jwallwork23 commented:
Okay I think this is ready for review. Apologies that it is a pretty big PR.

Presumably we want to keep any demo that will appear on the website fairly short, so I thought it might be useful to avoid duplicating `to_latlon` code. I propose to add a `COORDSYS` attribute in the `thetis.coordsys` model that can be tweaked to provide a general `to_latlon` function.

jwallwork23 commented:
Note that the mesh, bathymetry and station locations are included in `examples/north_sea` and then symlinked to `demos/`.

jhill1 commented:
Will do! Thanks for the heads up.

On Thu, 10 Mar 2022 at 11:03, Joe Wallwork ***@***.***> wrote:

> @jhill1 <https://github.com/jhill1>: if you have any comments on qmesh
> stuff, that would be much appreciated.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/thetisproject/thetis/pull/303#issuecomment-1063932735>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDFJKKO7FS6HM3NLFNCRGDU7HJJVANCNFSM5PXLLWFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Dr Jon Hill
Senior Lecturer in Physical Geography
Chair of Dept. Equality and Diversity Committee
Chair of Board of Examiners
Department of Environment and Geography
University of York
M: +44(0)7748254812
P: +44(0)1904 324480
Web: https://jonxhill.wordpress.com/
Web: https://envmodellinggroup.github.io/

<https://envmodellinggroup.github.io/>

AlexandrosAvdis reviewed (commented):
From my side, the comments in the mesh generation function could be slightly improved, but otherwise, the routine looks fine. The present terseness is rather attractive and justified as you are not trying to do a mesh generation (or qmesh) tutorial. This is a thetis example, and thetis is agnostic of how the mesh was generated.

However, line 52 ("Generate GMSH file") is slightly inaccurate, as you are not just generating gmsh files. Gmsh is also being invoked at that point, so "Generate mesh" might be a better option. Line 59 "Convert mesh coordinates" is rather misleading, as you are not converting coordinates, but you are converting the file type. You are reading the gmsh mesh and writing the mesh as a shapefile - for visualisation purposes. Therefore, the comment "Write mesh in shapefile format for visualisation in GIS software", might be more appropriate.

So mostly nitpicking from my side, alongside the admission that perhaps I should bear part of the blame as I am probably the partial originator of the incorrect comments!

stephankramer reviewed (commented):
This looks great! Maybe we should aim for a slightly better match that we can show, even if it's a coarse model.

Finally, I couldn't help but noticing there's some issues with the representation of the Netherlands :-) The IJsselmeer, the big lake in the middle which now looks like an open bay is in fact closed off and not tidally connected, so should not be meshed in. Then the area North of that, the Wadden Sea (behind the missing islands) is a bit tricky as it's a large intertidal flat which we're probably not going to do well in - not all that important if we're aiming for tides around the UK coast, but maybe we shouldn't include the Helgoland tidal gauge then
There's also some nasty mesh fingering along the Kattegat open boundary.

jwallwork23 commented:
Thanks very much @AlexandrosAvdis and @stephankramer for your really useful comments! I have pushed fixes for some items, but still need to tweak the domain geometry, simplify the coordsys stuff and try dropping viscosity.

jwallwork23 commented:
Also I will add some gradation to the ocean boundaries to try and avoid the mesh fingering near Kattegat.

jwallwork23 commented on demos/demo_2d_north_sea.py:
```diff
@@ -0,0 +1,239 @@
+
+viscosity_2d = Function(P1_2d, name="Viscosity coefficient")
+viscosity_2d.assign(0.5)
+
```
Indeed, I reran without and it made little difference. Will drop, thanks!

jwallwork23 commented on thetis/coordsys.py:
```diff
@@ -13,6 +13,7 @@
     datum='WGS84',
     units='m',
     errcheck=True)
+COORDSYS = UTM_ZONE10
```
Good points. I made a start on changing to providing the coordinate system, rather than a `to_latlon` function. However, some of the forcing/interpolation functions now end up getting passed this twice, so they will need redesigning a bit.

tkarna commented on examples/north_sea/README.md:
```diff
@@ -0,0 +1,110 @@
+[CMEMS catalogue](https://resources.marine.copernicus.eu/login), e.g. `IR`,
+`GL`, `NO`. Tide gauge data for the time period of interest were downloaded
+from the CMEMS dashboard and saved in the `data` subdirectory, so that they can
+be loaded by `plot_elevation.py`. These files have not been included here for
+copyright reasons. Note that you will need to register before you are able to
+download data from CMEMS.
```
Actually we might be able to include the CMEMS data in the repository, see the license in the annex section 2 [here](https://marine.copernicus.eu/user-corner/service-commitments-and-licence).

jwallwork23 replied:
Cool, I have now added it in an `observations` subdir. There is a page [here](https://help.marine.copernicus.eu/en/articles/4444611-how-to-cite-or-reference-copernicus-marine-products-and-services) about citing CMEMS, but I haven't yet been able to find out what the DOI would be for these in-situ observations.

jwallwork23 replied:
Found it! (on https://resources.marine.copernicus.eu/product-detail/INSITU_NWS_NRT_OBSERVATIONS_013_036/INFORMATION). Will add to demo and README.

tkarna commented on examples/north_sea/model_config.py:
```diff
@@ -0,0 +1,192 @@
+sim_tz = timezone.pytz.utc
+UTM_ZONE30 = pyproj.Proj(proj="utm", zone=30, datum="WGS84", units="m", errcheck=True)
+
+# Load tide gauge metadata
+with open("stations_elev.csv", "r") as csvfile:
+    stations = {
+        d["name"]: {
+            "latlon": (float(d["latitude"]), float(d["longitude"])),
+            "region": d["region"],
+        }
+        for d in csv.DictReader(csvfile, delimiter=",", skipinitialspace=True)
+    }
```
This is executed on import. Better to wrap it in a function? AFAIK it's only used in `construct_solver`

tkarna commented on examples/north_sea/model_config.py:
```diff
@@ -0,0 +1,192 @@
+    :kwarg cap: minimum value to cap the bathymetry at in the shallows
+    """
+    if cap <= 0.0:
+        raise NotImplementedError("Wetting and drying is not enabled in this example")
```
cap must be positive because ...

tkarna commented on examples/north_sea/spinup.py:
```diff
@@ -0,0 +1,49 @@
+mesh2d = solver_obj.mesh2d
+elev_init = Function(solver_obj.function_spaces.P1_2d)
+print_output(f"Exporting to {output_dir}")
+solver_obj.assign_initial_conditions(elev=elev_init)
```
There's no need to call `assign_initial_conditions` if `elev_init` is zero

tkarna commented on examples/north_sea/spinup.py:
```diff
@@ -0,0 +1,49 @@
+if os.getenv("THETIS_REGRESSION_TEST") is not None:
+    sys.exit(0)
+
+# Store spun-up state
+h5_dir = create_directory(os.path.join(output_dir, "hdf5"))
+exporter.HDF5Exporter(solver_obj.function_spaces.U_2d, h5_dir, "Velocity2d").export(
+    solver_obj.fields.uv_2d
+)
+exporter.HDF5Exporter(solver_obj.function_spaces.H_2d, h5_dir, "Elevation2d").export(
+    solver_obj.fields.elev_2d
+)
```
I would prefer to use normal exports to create the spun-up state, e.g.
```
fields_to_export = []
fields_to_export_hdf5 = ['elev_2d', 'uv_2d']
export_time = 24 * 3600.
```
Then to start the actual run you'd need to do `load_state(15, ..., t=0, iteration=0)` to reset the clock.
`HDF5Exporter` objects should only be used internally, so I would not include their usage in an example.

tkarna commented on examples/north_sea/generate_mesh.py:
```diff
@@ -0,0 +1,66 @@
+
+    # Read shapefile describing domain boundaries
+    boundary = qmesh.vector.Shapes()
+    boundary.fromFile("data/Boundary.shp")
```
Maybe the mesh generation related files could go to a separate subdir to differentiate from forcing data.

tkarna reviewed (changes requested):
Thanks a lot for this example! I only have minor comments.

We have not used symlinks in the repository before. Is that safe? What happens e.g. in windows? Maybe it does not matter as it's only used in the demo.

tkarna commented on demos/demo_2d_north_sea.py:
```diff
@@ -0,0 +1,234 @@
+# the source code for details on this. After a two week spin-up period, we obtain
+# the following free surface elevation field (as well as a velocity field).
+#
+# .. figure:: north_sea_init.png
```
I'd prefer a diverging symmetric colormap for the tides :)

jwallwork23 replied:
Is there one that you would recommend?

tkarna replied:
`"RdBu_r"` is usually ok.

tkarna commented on demos/demo_2d_north_sea.py:
```diff
@@ -0,0 +1,234 @@
+
+# The resulting bathymetry field is plotted below.
+#
+# .. figure:: north_sea_bathymetry.png
```
Maybe you could use a log scale colormap in this case or set the bounds to [0, 500] to show the shallow regions.

jwallwork23 commented on thetis/coordsys.py:
```diff
@@ -106,3 +106,18 @@ def __call__(self, v_x, v_y, i_node=None):
         return u, v
+
+
+def to_latlon(coord_system, x, y, positive_lon=False):
+    """
+    Convert model coordinates to latitude-longitude coordinates.
+
+    :arg coord_system: the local mesh coordinate system
+    :arg x: x coordinate
+    :arg y: y coordinate
+    :kwarg positive_lon: should positive longitude be enforced?
+    """
+    lon, lat = convert_coords(coord_system, LL_WGS84, x, y)
+    if positive_lon:
+        lon = numpy.mod(lon, 360.0)
+    return lat, lon
```
Ah okay, I see. I think I'll leave that for another PR if that's okay. I can't take credit for the log modulo fix - that was @stephankramer :)

jwallwork23 commented on demos/demo_2d_north_sea.py:
```diff
@@ -0,0 +1,234 @@
+#
+# In this demo, we are working with geographic data and so need to import a number of
+# additional packages and configure for the right timezones and map projection. The
+# domain of interest will be focused on the United Kingdom, so the timezone is UTC.
```
Oh okay. Do you think I should add to the text?

tkarna replied:
Probably not needed, just say that we  use UTC

jwallwork23 commented:
Thanks very much @tkarna for the detailed review! I have addressed most of your comments and asked follow-up questions on a couple of others. I also added citations more clearly in the demo.

Regarding the symlinking, I mainly used it to avoid adding the mesh twice. Open to alternative suggestions, but I think it should be fine for the purposes of a demo.

tkarna commented on examples/north_sea/README.md:
```diff
@@ -0,0 +1,110 @@
+exact coordinates - some have been modified slightly to ensure the gauges fall
+within the meshed domain.) We also include the relevant region codes used in the
+[CMEMS catalogue](https://resources.marine.copernicus.eu/login), e.g. `IR`,
+`GL`, `NO`. Tide gauge data for the time period of interest were downloaded
+from the CMEMS dashboard and saved in the `data` subdirectory, so that they can
+be loaded by `plot_elevation.py`. These files have not been included here for
```
`plot_elevation.py` seeks files in `forcing_data` subdirectory (a bit misleading name as we are not forcing the model with the observation data).

tkarna commented on examples/north_sea/plot_elevation.py:
```diff
@@ -0,0 +1,55 @@
+        vals_obs = nc.variables["SLEV"][:]
+        time_obs = nc.variables["TIME"][:]
+    vals_obs -= numpy.mean(vals_obs)
+    time_obs -= time_obs[0]
```
Yeah, I think this is wrong. The first time stamp in the monthly 202201 observation files stands for 2022-01-01T00 but the first time stamp in the model output files is 0 s simulation time i.e. 2022-01-15T00.

jwallwork23 replied:
Good spot! I fixed this and the results look much better.

tkarna commented on examples/north_sea/plot_setup.py:
```diff
@@ -0,0 +1,59 @@
+# Plot bathymetry in a logarithmic colourmap
+P1_2d = get_functionspace(mesh2d, "CG", 1)
+bathymetry2d = Function(P1_2d)
+interpolate_bathymetry(bathymetry2d)
```
This needs the `etopo1.nc` file which is not included in the repository. So the user cannot excecute this script. Maybe just plot the included `north_sea_bathymetry.h5` data instead?

jwallwork23 replied:
Good point

tkarna commented on examples/north_sea/plot_setup.py:
```diff
@@ -0,0 +1,60 @@
+    h5.load(bathymetry_2d)
+fig, axes = plt.subplots(figsize=(6.4, 4.8))
+triplot(mesh2d, axes=axes, boundary_kw={"linewidth": 0.5, "edgecolor": "k"})
+bathymetry2d.dat.data[:] = numpy.log10(bathymetry2d.dat.data[:])
+tc = tricontourf(bathymetry2d, axes=axes, levels=50)
```
bathymetry2d -> bathymetry_2d

jwallwork23 replied:
Ah yes, thanks.

tkarna commented on examples/north_sea/model_config.py:
```diff
@@ -0,0 +1,196 @@
+            solver_obj.add_callback(cb)
+
+    # Setup forcings
+    data_dir = os.path.join(os.environ.get("DATA", "~/data"), "tpxo")
```
This still uses `~/data` instead of `./data`
